frankenphp.c

```diff
@@ -583,92 +598,83 @@ static void frankenphp_register_known_variable(const char *key, char *value,
+                                     track_vars_array);
   frankenphp_register_known_variable("REMOTE_HOST", known_variables[9],
-                                     track_vars_array, true);
+                                     track_vars_array);
   frankenphp_register_known_variable("REMOTE_PORT", known_variables[10],
-                                     track_vars_array, true);
+                                     track_vars_array);
   frankenphp_register_known_variable("REQUEST_SCHEME", known_variables[11],
-                                     track_vars_array, true);
+                                     track_vars_array);
   frankenphp_register_known_variable("SCRIPT_FILENAME", known_variables[12],
-                                     track_vars_array, true);
+                                     track_vars_array);
   frankenphp_register_known_variable("SCRIPT_NAME", known_variables[13],
-                                     track_vars_array, true);
+                                     track_vars_array);
   frankenphp_register_known_variable("SERVER_NAME", known_variables[14],
-                                     track_vars_array, true);
+                                     track_vars_array);
   frankenphp_register_known_variable("SERVER_PORT", known_variables[15],
-                                     track_vars_array, true);
+                                     track_vars_array);
   frankenphp_register_known_variable("SERVER_PROTOCOL", known_variables[16],
-                                     track_vars_array, true);
+                                     track_vars_array);
   frankenphp_register_known_variable("SERVER_SOFTWARE", known_variables[17],
-                                     track_vars_array, true);
+                                     track_vars_array);
   frankenphp_register_known_variable("SSL_PROTOCOL", known_variables[18],
-                                     track_vars_array, true);
+                                     track_vars_array);
 
   size_t new_val_len;
-  for (size_t i = 0; i < size; i = i + 2) {
-    if (sapi_module.input_filter(
-            PARSE_SERVER, dynamic_variables[i], &dynamic_variables[i + 1],
-            strlen(dynamic_variables[i + 1]), &new_val_len)) {
-      php_register_variable_safe(dynamic_variables[i], dynamic_variables[i + 1],
-                                 new_val_len, track_vars_array);
+  for (size_t i = 0; i < size; i++) {
+    if (sapi_module.input_filter(PARSE_SERVER, dynamic_variables[i].var,
+                                 &dynamic_variables[i].data,
+                                 dynamic_variables[i].data_len, &new_val_len)) {
+      php_register_variable_safe(dynamic_variables[i].var,
+                                 dynamic_variables[i].data, new_val_len,
+                                 track_vars_array);
     }
-
-    free(dynamic_variables[i]);
-    free(dynamic_variables[i + 1]);
   }
+
+  free(dynamic_variables);
```

Comment:
Won't this just free `dynamic_variables[0]`? I haven't loaded this PR into an IDE yet, but I suspect if we are putting these into the otter cache, we'll result in double free or use-after-free? It appears we are storing dynamic vars in there?

Reply:
`free()` should free the whole block allocated by `malloc()`. Why would it free only the first element (I don't think that it's even possible to do so)?

Reply:
Ah, I didn't catch that the type changed from `char **` to `php_variable *`.

Should be fine then.

Reply:
Allocating on Go side (last commit) seems to fix the issue and remove a bit of unsafe trickery.

We'll likely keep it as-is but I'm very curious of what's going on with this call to `free()`. Maybe a bug in cgo?